pkg/services/ngalert/state/historian/otel_loki_client.go

```diff
@@ -0,0 +1,184 @@
+		exportMethodName = "OtelExport"
+	)
+	p.once.Do(func() {
+		var conn *grpc.ClientConn
+		conn, err = newOtlpGrpcConn(p.cfg)
+		if err != nil {
+			return
+		}
+		p.client = plogotlp.NewGRPCClient(conn)
+	})
+
+	if err != nil {
+		return err
+	}
+
+	logs, size, err := p.pushRequestToLogs(s, time.Now())
+	if err != nil {
+		return err
+	}
+
+	exportStart := time.Now()
+	p.metrics.WriteDuration.Before(ctx, exportMethodName, exportStart)
+	_, err = p.client.Export(ctx, plogotlp.NewExportRequestFromLogs(logs))
+	if err != nil {
+		return fmt.Errorf("failed to export logs: %w", err)
+	}
+	p.metrics.WriteDuration.After(ctx, exportMethodName, timerFailureCode, exportStart)
+	p.metrics.BytesWritten.Add(float64(size))
+	return nil
+}
+
+func (p *otelLokiClient) pushRequestToLogs(sreams []Stream, defaultTimestamp time.Time) (plog.Logs, int, error) {
+	logs := plog.NewLogs()
+	if len(sreams) == 0 {
+		return logs, 0, nil
+	}
+	rls := logs.ResourceLogs().AppendEmpty()
+	logSlice := rls.ScopeLogs().AppendEmpty().LogRecords()
+	totalSize := 0
+
+	var lastErr error
+	var errNumber int64
+	for _, stream := range sreams {
+		// Return early if stream does not contain any entries
+		if len(stream.Stream) == 0 {
+			continue
+		}
+
+		totalSize += calcAttributesSize(stream.Stream)
+
+		for _, entry := range stream.Values {
+			lr := logSlice.AppendEmpty()
+			convertEntryToLogRecord(entry, stream.Stream, &lr, defaultTimestamp)
+			totalSize += len(entry.V)
+		}
+	}
+
+	if lastErr != nil {
+		lastErr = fmt.Errorf("%d entries failed to process, the last error: %w", errNumber, lastErr)
+	}
+
+	return logs, totalSize, lastErr
+}
+
+func convertEntryToLogRecord(entry Sample, streamAttributes map[string]string, lr *plog.LogRecord, defaultTimestamp time.Time) error {
+	const timestampAttribute = "timestamp"
+
+	observedTimestamp := pcommon.NewTimestampFromTime(defaultTimestamp)
+	lr.SetObservedTimestamp(observedTimestamp)
+
+	var recordAttributes map[string]any
+	err := json.Unmarshal([]byte(entry.V), &recordAttributes)
+	if err != nil {
+		return fmt.Errorf("failed to unmarshal log line: %w", err)
+	}
+
+	var timestamp pcommon.Timestamp
+	if !entry.T.IsZero() {
+		timestamp = pcommon.NewTimestampFromTime(entry.T)
+	} else {
+		timestamp = observedTimestamp
+	}
+
+	lr.SetTimestamp(timestamp)
+	recordAttributes[timestampAttribute] = timestamp.AsTime().Format(time.RFC3339Nano)
+
+	for k, v := range streamAttributes {
+		recordAttributes[k] = v
+	}
+
+	lr.Attributes().FromRaw(recordAttributes)
+	return nil
+}
+
+func calcAttributesSize(attributes map[string]string) int {
+	size := 0
+	for k, v := range attributes {
+		size += len(k) + len(v)
+	}
+	return size
+}
+
+func newOtlpGrpcConn(cfg OtelConfig) (conn *grpc.ClientConn, err error) {
+	const (
+		apiKeyHeader                 = "apikey"
+		defaultConnectionDialTimeout = 10 * time.Second
+	)
+	creds := insecure.NewCredentials()
+	if cfg.EnableTLS {
+		config := &tls.Config{
+			InsecureSkipVerify: cfg.TLSSkipVerify,
```

Comment:
right now we support forcing verify on monitors? we had that issue with alligator in a client recently